feat: add `lapack/base/dgttrf` (#5754)
stdlib-js/stdlib: lib/node_modules/@stdlib/lapack/base/dgttrf/README.md

<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# dgttrf

> > Compute an `LU` factorization of a real tri diagonal matrix `A` using elimination with partial pivoting and row interchanges.

Reviewer: ```suggestion
> Compute an `LU` factorization of a real tridiagonal matrix `A` using elimination with partial pivoting and row interchanges.
```

Here and elsewhere.

Reviewer: Looks like this was never resolved. Need to update `tri diagonal` to `tridiagonal` here and elsewhere.

> 

Reviewer: Another option is to write an introduction which includes TeX equations to illustrate what this package does.

> Given the following square matrix `A`,

(insert equation of matrix)

> the `LU` factorization is defined as

(insert equation of LU)

> where `L` is ... and `U` is ...

and then can we describe in terms of the overwritten arrays below?

<section class="usage">

## Usage

```javascript
var dgttrf = require( '@stdlib/lapack/base/dgttrf' );
```

#### dgttrf( N, DL, D, DU, DU2, IPIV )

Computes an `LU` factorization of a real tri diagonal matrix `A` using elimination with partial pivoting and row interchanges.

```javascript
var Float64Array = require( '@stdlib/array/float64' );
var Int32Array = require( '@stdlib/array/int32' );

> 

Reviewer: In order to help visualize the matrix `A`, would be good to add a comment with a visual illustration so that we know how `A` breaks down into its constituent parts below.

Reviewer: ```suggestion
/*
A = [
           2.0, 1.0, 0.0,
           1.0, 3.0, 1.0,
           0.0, 1.0, 1.0
      ]
*/
```

Reviewer: I'd also suggest making the values array "specific". E.g., upper diagonal, use `5.0, 6.0` and lower diagonal `8.0, 9.0`.

Also to expand to include second diagonal, make a 4x4 matrix so you can use a second super diagonal having values other than `0`.

var DL = new Float64Array( [ 1.0, 1.0 ] );
var D = new Float64Array( [ 2.0, 3.0, 1.0 ] );
var DU = new Float64Array( [ 1.0, 1.0 ] );
var DU2 = new Float64Array( 1 );
var IPIV = new Int32Array( 3 );

dgttrf( 3, DL, D, DU, DU2, IPIV );
// DL => <Float64Array>[ 0.5, 0.4 ]
// D => <Float64Array>[ 2, 2.5, 0.6 ]
// DU => <Float64Array>[ 1, 1 ]
// DU2 => <Float64Array>[ 0 ]
// IPIV => <Int32Array>[ 0, 1, 2 ]
```

The function has the following parameters:

> -   **N**: order of matrix `A`.

Reviewer: ```suggestion
-   **N**: order of matrix `A`.
```

I suggest updating this description to avoid confusion with "order" as the memory layout. Here, we mean "rank" or the number of rows/columns in a square matrix `A`.

> -   **DL**: the sub diagonal elements of `A` as a [`Float64Array`][mdn-float64array]. On exit, DL is overwritten by the multipliers that define the matrix `L` from the `LU` factorization of `A`.

Reviewer: ```suggestion
-   **DL**: the sub diagonal elements of `A` as a [`Float64Array`][mdn-float64array]. `DL` is overwritten by the multipliers that define the matrix `L` from the `LU` factorization of `A`.
```

> -   **DL**: the sub diagonal elements of `A` as a [`Float64Array`][mdn-float64array]. On exit, DL is overwritten by the multipliers that define the matrix `L` from the `LU` factorization of `A`.

Reviewer: I would also document the expected number of indexed elements of this array.

> -   **DL**: the sub diagonall elements of `A` as a [`Float64Array`][mdn-float64array]. On exit, DL is overwritten by the multipliers that define the matrix `L` from the `LU` factorization of `A`.

Reviewer: ```suggestion
-   **DL**: the sub diagonal elements of `A` as a [`Float64Array`][mdn-float64array]. On exit, DL is overwritten by the multipliers that define the matrix `L` from the `LU` factorization of `A`.
```

> -   **D**: the diagonal elements of `A` as a [`Float64Array`][mdn-float64array]. On exit, D is overwritten by the diagonal elements of the upper triangular matrix `U` from the `LU` factorization of `A`.

Reviewer: You can remove the phrase `On exit`.

> -   **DU**: the super diagonal elements of `A` as a [`Float64Array`][mdn-float64array]. On exit, DU is overwritten by the elements of the first super-diagonal of `U`.

Reviewer: ```suggestion
-   **DU**: the first super-diagonal elements of `A` as a [`Float64Array`][mdn-float64array]. `DU` is overwritten by the elements of the first super-diagonal of `U`.
```

Reviewer: Also need to ensure that consistent with `super-diagonal` spelling.

> -   **DU2**: On exit, DU2 is overwritten by the elements of the second super-diagonal of `U` as a [`Float64Array`][mdn-float64array].

Reviewer: ```suggestion
-   **DU2**: the second super diagonal of `U`  as a [`Float64Array`][mdn-float64array]. `DU2` is overwritten by the elements of the second super-diagonal of `U`.
```

-   **IPIV**: vector of pivot indices as a [`Int32Array`][mdn-int32array].

Note that indexing is relative to the first index. To introduce an offset, use [`typed array`][mdn-typed-array] views.

<!-- eslint-disable stdlib/capitalized-comments -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );
var Int32Array = require( '@stdlib/array/int32' );

// Initial arrays...
var DL0 = new Float64Array( [ 0.0, 1.0, 1.0 ] );
var D0 = new Float64Array( [ 0.0, 2.0, 3.0, 1.0 ] );
var DU0 = new Float64Array( [ 0.0, 1.0, 1.0 ] );
var DU20 = new Float64Array( 2 );
var IPIV0 = new Int32Array( 4 );

// Create offset views...
var DL = new Float64Array( DL0.buffer, DL0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
var D = new Float64Array( D0.buffer, D0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
var DU = new Float64Array( DU0.buffer, DU0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
var DU2 = new Float64Array( DU20.buffer, DU20.BYTES_PER_ELEMENT*1 ); // start at 2nd element
var IPIV = new Int32Array( IPIV0.buffer, IPIV0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

dgttrf( 3, DL, D, DU, DU2, IPIV );

> // DL0 => <Float64Array>[ 0, 0.5, 0.4 ]

Reviewer: ```suggestion
// DL0 => <Float64Array>[ 0.0, 0.5, 0.4 ]
```

Want to add decimals here and below for the `Float64Array` values.

// D0 => <Float64Array>[ 0, 2, 2.5, 0.6 ]
// DU0 => <Float64Array>[ 0, 1, 1 ]
// DU20 => <Float64Array>[ 0, 0 ]
// IPIV0 => <Int32Array>[ 0, 0, 1, 2 ]
```

<!-- lint disable maximum-heading-length -->

#### dgttrf.ndarray( N, DL, sdl, odl, D, sd, od, DU, sdu, odu, DU2, sdu2, odu2, IPIV, si, oi )

Computes an `LU` factorization of a real tri diagonal matrix `A` using elimination with partial pivoting and row interchanges and alternative indexing semantics.

```javascript
var Float64Array = require( '@stdlib/array/float64' );
var Int32Array = require( '@stdlib/array/int32' );
var dgttrf = require( '@stdlib/lapack/base/dgttrf' );

var DL = new Float64Array( [ 1.0, 1.0 ] );
var D = new Float64Array( [ 2.0, 3.0, 1.0 ] );
var DU = new Float64Array( [ 1.0, 1.0 ] );
var DU2 = new Float64Array( 1 );
var IPIV = new Int32Array( 3 );

dgttrf.ndarray( 3, DL, 1, 0, D, 1, 0, DU, 1, 0, DU2, 1, 0, IPIV, 1, 0 );

> // DL => <Float64Array>[ 0.5, 0.4 ]

Reviewer: Decimals for `Float64Array` values.

// D => <Float64Array>[ 2, 2.5, 0.6 ]
// DU => <Float64Array>[ 1, 1 ]
// DU2 => <Float64Array>[ 0 ]
// IPIV => <Int32Array>[ 0, 1, 2 ]
```

The function has the following additional parameters:

-   **sdl**: stride length for `DL`.
-   **odl**:  starting index for `DL`.
-   **sd**: stride length for `D`.
-   **od**:  starting index for `D`.
-   **sdu**: stride length for `DU`.
-   **odu**:  starting index for `DU`.
-   **sdu2**: stride length for `DU2`.
-   **odu2**:  starting index for `DU2`.
-   **si**: stride length for `IPIV`.
-   **oi**:  starting index for `IPIV`.

While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example,

<!-- eslint-disable max-len -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );
var Int32Array = require( '@stdlib/array/int32' );

var DL = new Float64Array( [ 0.0, 1.0, 1.0 ] );
var D = new Float64Array( [ 0.0, 2.0, 3.0, 1.0 ] );
var DU = new Float64Array( [ 0.0, 1.0, 1.0 ] );

> var DU2 = new Float64Array( 2 );

Reviewer: ```suggestion
var DU2 = new Float64Array( [ 0.0, 0.0 ] );
```

> var IPIV = new Int32Array( 4 );

Reviewer: ```suggestion
var IPIV = new Int32Array( [ 0, 0, 0, 0 ] );
```

> dgttrf.ndarray( 3, DL, 1, 1, D, 1, 1, DU, 1, 1, DU2, 1, 1, IPIV, 1, 1 );

Reviewer: Same comment about examples.

// DL => <Float64Array>[ 0, 0.5, 0.4 ]
// D => <Float64Array>[ 0, 2, 2.5, 0.6 ]
// DU => <Float64Array>[ 0, 1, 1 ]
// DU2 => <Float64Array>[ 0, 0 ]
// IPIV => <Int32Array>[ 0, 0, 1, 2 ]
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

> -   Both functions mutate the input arrays `DL`, `D`, `DU`, `DU2` and `IPIV`.

Reviewer: ```suggestion
-   Both functions mutate the input arrays `DL`, `D`, `DU`, `DU2`, and `IPIV`.
```

-   Both functions return a status code indicating success or failure. A status code indicates the following conditions:

    -   `0`: factorization was successful.

>     -   `<0`: the k-th argument had an illegal value, where `-k` equals the status code value.

Reviewer: You can remove this status code condition, as we never reach.

    -   `>0`: `U( k, k )` is exactly zero the factorization has been completed, but the factor `U` is exactly singular, and division by zero will occur if it is used to solve a system of equations, where `k` equals the status code value.

-   `dgttrf()` corresponds to the [LAPACK][LAPACK] routine [`dgttrf`][lapack-dgttrf].

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Int32Array = require( '@stdlib/array/int32' );
var Float64Array = require( '@stdlib/array/float64' );
var dgttrf = require( '@stdlib/lapack/base/dgttrf' );

> var N = 9;

Reviewer: Would be good to "show" the matrix `A`.

<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# dgttrf

> Compute an `LU` factorization of a real tridiagonal matrix `A` using elimination with partial pivoting and row interchanges.

<section class="intro">

> The `dgttrf` routine computes an LU factorization of a real n-by-n tridiagonal matrix A using elimination with partial pivoting and row interchanges. The factorization has the form:

Reviewer: ```suggestion
The `dgttrf` routine computes an LU factorization of a real n-by-n tridiagonal matrix `A` using elimination with partial pivoting and row interchanges. The factorization has the form:
```

<!-- <equation class="equation" label="eq:lu_decomposition" align="center" raw="A = L U" alt="Equation for LU factorization."> -->

```math
A = L U
```

<!-- <div class="equation" align="center" data-raw-text="A = L U" data-equation="eq:lu_decomposition"></div> -->

<!-- </equation> -->

> where L is a product of permutation and unit lower bidiagonal matrices and U is upper triangular with nonzeros in only the main diagonal and first two superdiagonals.

Reviewer: ```suggestion
where `L` is a product of permutation and unit lower bidiagonal matrices and `U` is upper triangular with nonzeros in only the main diagonal and first two superdiagonals.
```

> For a tridiagonal matrix A, its elements are stored in three arrays:

Reviewer: ```suggestion
For a tridiagonal matrix `A`, its elements are stored in three arrays:
```

> <!-- <equation class="equation" label="eq:matrix_a" align="center" raw="A = \begin{bmatrix}

Reviewer: For these equations and below, I suggest consulting the Wikipedia page for banded matrices (https://en.wikipedia.org/wiki/Band_matrix) and matching the TeX representation wrt matrix representation (e.g., use of ellipses and when to show a zero). Overall, these look quite nice. Thanks for adding.

    d_1    & du_1   & 0      & \cdots & 0      \\
    dl_1   & d_2    & du_2   & \cdots & 0      \\
    0      & dl_2   & d_3    & \ddots & \vdots \\
    \vdots & \ddots & \ddots & \ddots & du_{n-1}\\
    0      & \cdots & 0      & dl_{n-1} & d_n
\end{bmatrix}" alt="Representation of matrix A."> -->

```math
A = \begin{bmatrix}
    d_1    & du_1   & 0      & \cdots & 0      \\
    dl_1   & d_2    & du_2   & \cdots & 0      \\
    0      & dl_2   & d_3    & \ddots & \vdots \\
    \vdots & \ddots & \ddots & \ddots & du_{n-1}\\
    0      & \cdots & 0      & dl_{n-1} & d_n
\end{bmatrix}
```

<!-- <div class="equation" align="center" data-raw-text="A = \begin{bmatrix}
    d_1    & du_1   & 0      & \cdots & 0      \\
    dl_1   & d_2    & du_2   & \cdots & 0      \\
    0      & dl_2   & d_3    & \ddots & \vdots \\
    \vdots & \ddots & \ddots & \ddots & du_{n-1}\\
    0      & \cdots & 0      & dl_{n-1} & d_n
\end{bmatrix}" data-equation="eq:matrix_a"></div> -->

<!-- </equation> -->

where:

> -   `dl` contains the subdiagonal elements
> -   `d` contains the diagonal elements
> -   `du` contains the superdiagonal elements

Reviewer: ```suggestion
-   `dl` contains the subdiagonal elements.
-   `d` contains the diagonal elements.
-   `du` contains the superdiagonal elements.
```

> After factorization, the elements of L and U overwrite the input arrays, where:

Reviewer: ```suggestion
After factorization, the elements of `L` and `U` overwrite the input arrays, where:
```

-   `dl` contains the multipliers that define unit lower bidiagonal matrix L
-   `d` contains the diagonal elements of U

> -   `du` and `du2` contain the elements of U on the first and second superdiagonals

Reviewer: ```suggestion
-   `dl` contains the multipliers that define unit lower bidiagonal matrix `L`.
-   `d` contains the diagonal elements of `U`.
-   `du` and `du2` contain the elements of `U` on the first and second superdiagonals.
```

> The resulting L and U matrices have the following structure:

Reviewer: ```suggestion
The resulting `L` and `U` matrices have the following structure:
```
